expose parse_schema and __eq__ function to python and add round trip tests

houseroad opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23208 expose parse_schema and __eq__ function to python and add round trip tests**

expose necessary functions to python, and add round-way tests for
function schema str() and parsing functions.
We iterate over all the registered function schemas and get the string,
then parse the string. We compare the schema generated from parsing with
the original one, and make sure they are equal.

Differential Revision: [D16435471](https://our.internmc.facebook.com/intern/diff/D16435471/)

houseroad committed:
expose parse_schema and __eq__ function to python and add round trip tests

expose necessary functions to python, and add round-way tests for
function schema str() and parsing functions.
We iterate over all the registered function schemas and get the string,
then parse the string. We compare the schema generated from parsing with
the original one, and make sure they are equal.

Differential Revision: [D16435471](https://our.internmc.facebook.com/intern/diff/D16435471/)

houseroad committed:
Update on "expose parse_schema and __eq__ function to python and add round trip tests"


expose necessary functions to python, and add round-way tests for
function schema str() and parsing functions.
We iterate over all the registered function schemas and get the string,
then parse the string. We compare the schema generated from parsing with
the original one, and make sure they are equal.

Differential Revision: [D16435471](https://our.internmc.facebook.com/intern/diff/D16435471/)

houseroad committed:
Update on "expose parse_schema and __eq__ function to python and add round trip tests"


expose necessary functions to python, and add round-way tests for
function schema str() and parsing functions.
We iterate over all the registered function schemas and get the string,
then parse the string. We compare the schema generated from parsing with
the original one, and make sure they are equal.

Differential Revision: [D16435471](https://our.internmc.facebook.com/intern/diff/D16435471/)

houseroad committed:
Update on "expose parse_schema and __eq__ function to python and add round trip tests"


expose necessary functions to python, and add round-way tests for
function schema str() and parsing functions.
We iterate over all the registered function schemas and get the string,
then parse the string. We compare the schema generated from parsing with
the original one, and make sure they are equal.

Differential Revision: [D16435471](https://our.internmc.facebook.com/intern/diff/D16435471/)

dzhulgakov commented on test/test_function_schema.py:
```diff
@@ -0,0 +1,18 @@
+from common_utils import TestCase
+
+
+class TestSparse(TestCase):
```
change name?

dzhulgakov commented on test/test_function_schema.py:
```diff
@@ -0,0 +1,18 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
```
if you're creating a new file - we should put it into TARGETS and some run_test.py files I think. Otherwise it won't be run in CI. Or just dump it into test_jit.py :)

houseroad replied:
I have put it into TARGETS, forgot to add it to run_test.py. Good catch!

houseroad committed:
Update on "expose parse_schema and __eq__ function to python and add round trip tests"


expose necessary functions to python, and add round-way tests for
function schema str() and parsing functions.
We iterate over all the registered function schemas and get the string,
then parse the string. We compare the schema generated from parsing with
the original one, and make sure they are equal.

Differential Revision: [D16435471](https://our.internmc.facebook.com/intern/diff/D16435471/)

dzhulgakov reviewed (approved):
Looks like you need to fix the doctest here (please do before landing): https://github.com/pytorch/pytorch/blob/8c927b208ca9ca5b8bac481f9b54d104038def3c/test/test_docs_coverage.py#L46

houseroad committed:
Update on "expose parse_schema and __eq__ function to python and add round trip tests"


expose necessary functions to python, and add round-way tests for
function schema str() and parsing functions.
We iterate over all the registered function schemas and get the string,
then parse the string. We compare the schema generated from parsing with
the original one, and make sure they are equal.

Differential Revision: [D16435471](https://our.internmc.facebook.com/intern/diff/D16435471/)

houseroad committed:
Update on "expose parse_schema and __eq__ function to python and add round trip tests"


expose necessary functions to python, and add round-way tests for
function schema str() and parsing functions.
We iterate over all the registered function schemas and get the string,
then parse the string. We compare the schema generated from parsing with
the original one, and make sure they are equal.

Differential Revision: [D16435471](https://our.internmc.facebook.com/intern/diff/D16435471/)

houseroad committed:
Update on "expose parse_schema and __eq__ function to python and add round trip tests"


expose necessary functions to python, and add round-way tests for
function schema str() and parsing functions.
We iterate over all the registered function schemas and get the string,
then parse the string. We compare the schema generated from parsing with
the original one, and make sure they are equal.

Differential Revision: [D16435471](https://our.internmc.facebook.com/intern/diff/D16435471/)

houseroad committed:
Update on "expose parse_schema and __eq__ function to python and add round trip tests"


expose necessary functions to python, and add round-way tests for
function schema str() and parsing functions.
We iterate over all the registered function schemas and get the string,
then parse the string. We compare the schema generated from parsing with
the original one, and make sure they are equal.

Differential Revision: [D16435471](https://our.internmc.facebook.com/intern/diff/D16435471/)

houseroad committed:
Update on "expose parse_schema and __eq__ function to python and add round trip tests"


expose necessary functions to python, and add round-way tests for
function schema str() and parsing functions.
We iterate over all the registered function schemas and get the string,
then parse the string. We compare the schema generated from parsing with
the original one, and make sure they are equal.

Differential Revision: [D16435471](https://our.internmc.facebook.com/intern/diff/D16435471/)
